### PR TITLE
Make an example for seq2seq

### LINKED ISSUE #2060: Make seq2seq example using NStepLSTM
We need to make seq2seq example.

### PR DESCRIPTION
**Pull request by unnonouno**
I made an example of sequence-to-sequence learning for machine translation.

fix #2060

### CODE REVIEW

File: examples/seq2seq/seq2seq.py [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+        exs = sequence_embed(self.embed_x, xs)
+        eys = sequence_embed(self.embed_y, ys_in)
+
+        batch = len(xs)
+        # Initial hidden variable and cell variable
+        zero = self.xp.zeros((self.n_layers, batch, self.n_units), 'f')
+        hx, cx, _ = self.encoder(zero, zero, exs)
+        _, _, os = self.decoder(hx, cx, eys)
+        loss = F.softmax_cross_entropy(
+            self.W(F.concat(os, axis=0)), F.concat(ys_out, axis=0))
```
Comment (soskek) [inline examples/seq2seq/seq2seq.py]: You simply ignore the variation of normalization coefficients to devide in `softmax_cross_entropy` for each batch, don't you? A typical coefficient is the batchsize (= `len(ys_out)`), while this implementation uses the total numbers of whole tokens in a batch (= `sum(y.shape[0] for y in ys_out)`) for it.
(If you already recognized it and do so for simplicity, that's fine.)

File: examples/seq2seq/seq2seq.py [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+        zero = self.xp.zeros((self.n_layers, 1, self.n_units), 'f')
+        h, c, _ = self.encoder(zero, zero, exs)
+        y = self.xp.zeros(1, 'i')
+        result = []
+        for i in range(10):
+            ey = self.embed_y(y)
+            h, c, ys = self.decoder(h, c, [ey])
+            wy = self.W(ys[0])
+            # TODO(unno): CuPy does not have argmax method
+            yi = numpy.argmax(cuda.to_cpu(wy.data[0]))
```
Comment (soskek) [inline examples/seq2seq/seq2seq.py]: Now, `cupy.argmax()` looks available too.

Comment (unnonouno) [inline examples/seq2seq/seq2seq.py]: OK

File: examples/seq2seq/seq2seq.py [hunk 3]
```diff
@@ -0,0 +1,173 @@
+import argparse
+import collections
+
+from nltk.corpus import comtrans
+from nltk.translate import bleu_score
```
Comment (unnonouno) [inline examples/seq2seq/seq2seq.py]: Yes, I'll write README for this example in this directory. Other examples also uses additional libraries.

File: examples/seq2seq/seq2seq.py (hunk 1 above)
Comment (unnonouno) [inline examples/seq2seq/seq2seq.py]: Ah, ok. i'll fix it.

File: examples/seq2seq/README.md [hunk 4]
```diff
@@ -0,0 +1,75 @@
+# Sequence-to-sequnce learning example for machine translation
+
+This is a minimal example of sequence-to-sequence learning. Sequence-to-sequence learning model is a conversion model from a sequence to a sequence. You can regard many tasks in natural language processing field as this type of task, such as machine translation, dialogue and summarization.
```
Comment (bkvogel) [inline examples/seq2seq/README.md]: Consider rewording:

This is a minimal example of sequence-to-sequence learning. Sequence-to-sequence is a learning model that converts an input sequence into an output sequence. You can regard many tasks in the natural language processing field as this type of task, such as machine translation, dialogue and summarization.

File: examples/seq2seq/README.md [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+This example sciprt is simplest one. An input sequence is processed by a stacked LSTM-RNN and it is encoded as a fixed-size of vector. And the an output sequence is also processed by another stacked LSTM-RNN. In decoding time, an output sequence is generated in argmax fashion.
```
Comment (bkvogel) [inline examples/seq2seq/README.md]: In this simple example script, an input sequence is processed by a stacked LSTM-RNN and it is encoded as a fixed-size vector. The output sequence is also processed by another stacked LSTM-RNN. At decoding time, an output sequence is generated using argmax.

File: examples/seq2seq/README.md [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+## Dataset format
+
+You need to prepare four files.
+
+1. Source language sentence file
+2. Source language vocabulary file
+3. Targe language sentence file
+4. Target language vocabulary file
+
+In sentence files each line represents a sentece. In each line, each sentence needs to be separated into words by space characters.
```
Comment (bkvogel) [inline examples/seq2seq/README.md]: In the sentence files, each line represents a sentence.

File: examples/seq2seq/README.md [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+Each line corresponds to other one. So both file need to have the same numbers of lines.
```
Comment (bkvogel) [inline examples/seq2seq/README.md]: Since the number of source and target sentences is the same, note that both files need to have the same number of lines.

**Comment by unnonouno** [comment]
Thank you. I fixed readme.

File: examples/seq2seq/README.md [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+$ tar zxf dev-v2.tgz
+$ python wmt_preprocess.py dev/newstest2013.en newstest2013.preprocess.en
+$ python wmt_preprocess.py dev/newstest2013.fr newstest2013.preprocess.fr
+```
+
+Let's start training. Add `--validation-source` and `--validation-target` argument
+to specify validation dataset.
+
+```
+$ python seq2seq.py --gpu=0 giga-fen.preprocess.en giga-fren.preprocess.fr \
```
Comment (himkt) [inline examples/seq2seq/README.md]: I think `giga-fen.preprocess.en` is typo for `giga-fren.preprocess.en`.

Comment (himkt) [inline examples/seq2seq/README.md]: I think giga-fen.preprocess.en is typo for giga-fren.preprocess.en.

Comment (unnonouno) [inline examples/seq2seq/README.md]: Oh, thank you

File: examples/seq2seq/seq2seq.py [hunk 9]
```diff
@@ -0,0 +1,302 @@
...
+                sources = [
+                    chainer.dataset.to_device(self.device, x) for x in sources]
+                ys = [y.tolist()
+                      for y in self.model.translate(sources, self.max_length)]
+                hypotheses.extend(ys)
+
+        bleu = bleu_score.corpus_bleu(
+            references, hypotheses,
+            smoothing_function=bleu_score.SmoothingFunction().method1)
+        reporter.report({self.key: bleu})
```
Comment (beam2d) [inline examples/seq2seq/seq2seq.py]: Use `chainer.report`

File: examples/seq2seq/seq2seq.py [hunk 10]
```diff
@@ -0,0 +1,302 @@
+#!/usr/bin/env python
+
+import argparse
+
+from nltk.translate import bleu_score
```
Comment (beam2d) [inline examples/seq2seq/seq2seq.py]: Mention in the readme that `NLTK` is required as a dependency.

File: examples/seq2seq/seq2seq.py [hunk 11]
```diff
@@ -0,0 +1,302 @@
...
+
+    optimizer = chainer.optimizers.Adam()
+    optimizer.setup(model)
+
+    train_iter = chainer.iterators.SerialIterator(train_data, args.batchsize)
+    updater = training.StandardUpdater(
+        train_iter, optimizer, converter=convert, device=args.gpu)
+    trainer = training.Trainer(updater, (args.epoch, 'epoch'))
+    trainer.extend(extensions.LogReport(trigger=(200, 'iteration')),
+                   trigger=(200, 'iteration'))
```
Comment (beam2d) [inline examples/seq2seq/seq2seq.py]: Is it correct? This setting will not take average training loss over 200 iterations.

File: examples/seq2seq/seq2seq.py [hunk 12]
```diff
@@ -0,0 +1,301 @@
...
+
+
+def count_lines(path):
+    with open(path) as f:
+        return sum([1 for _ in f])
+
+
+def load_vocabulary(path):
+    with open(path) as f:
+        return {line.strip(): i for i, line in enumerate(f)}
```
Comment (jun-harashima) [inline examples/seq2seq/seq2seq.py]: Because of `UNK` and `EOS`, `i` -> `i + 2` ?

File: examples/seq2seq/README.md [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+
+
+## Tarin with WMT dataset
+
+First you need to prepare parallel corpus. Download 10^9 French-English corpus from WMT15 website.
+
+http://www.statmt.org/wmt15/translation-task.html
+
+```
+$ http://www.statmt.org/wmt10/training-giga-fren.tar
```
Comment (himkt) [inline examples/seq2seq/README.md]: before: `http://www.statmt.org/wmt10/training-giga-fren.tar` 
after: `wget http://www.statmt.org/wmt10/training-giga-fren.tar`

File: examples/seq2seq/seq2seq.py (hunk 12 above)
Comment (unnonouno) [reply examples/seq2seq/seq2seq.py]: Thank you!

File: examples/seq2seq/README.md (hunk 13 above)
Comment (unnonouno) [reply examples/seq2seq/README.md]: Thank you!

**Comment by beam2d** [comment]
LGTM

**Comment by beam2d** [comment]
Thank you!
